patch/driver-support-optoe-non-sfp-reset-page-select.patch

```diff
@@ -0,0 +1,65 @@
+From: Mihir Patel <patelmi@microsoft.com>
+Date: Wed, 12 Feb 2025 21:04:48 +0000
+Subject: [PATCH] Reset page select byte to 0
+
+Description
+The optoe kernel driver currently assumes that the page select byte (page 0, byte 127) is set to 0h before accessing the upper memory on page 0h.
+This assumption causes failures when accessing the module's EEPROM in multiple scenarios.
+On the contrary, the driver currently sets the page select byte if the intended page to be accessed is > 0 as evident from the below code
+https://github.com/opencomputeproject/oom/blob/c32499a89dff005e7ff2acb48b33f55543ce5140/optoe/optoe.c#L344-L353
+
+Impact
+If the optoe driver accesses any page except page 0h and the code for restoring the page select byte to 0h fails, subsequent attempts to
+access the upper memory of page 0h will fail because the page select byte from the previous transaction remains unchanged.
+https://github.com/opencomputeproject/oom/blob/c32499a89dff005e7ff2acb48b33f55543ce5140/optoe/optoe.c#L373-L392
+
+Steps to recreate the failure scenario
+One way to recreate this issue is:
+
+1. Initiate CDB firmware download operation on a module supporting CDB foreground mode only
+2. Kill the process which is downloading the FW
+3. Dump the page select byte and ensure it displays a non-zero value
+4. Execute sfputil show fwversion EthernetXX CLI and you will observe a traceback. The traceback occurs because the vendor name is read
+with garbage content, as the page select byte is set to 0x9F instead of 0x0.
+
+Fix
+Before optoe accesses any address from upper page 0h, the page select byte is now set to 0.
+This fix is done only for non-SFP modules to limit the scope of impact. SFP modules require additional handling since the page select byte
+is on page A2h and NOT A0h unlike non-SFP modules which do not have the concept of A0h and A2h page.
+
+Testing
+The fix has been tested on a non-SFP module and it was ensured that the sfputil show fwversion EthernetXX CLI command executes successfully without any traceback.
+
+Signed-off-by: Mihir Patel <patelmi@microsoft.com>
+---
+ drivers/misc/eeprom/optoe.c | 13 +++++++++++++
+ 1 file changed, 13 insertions(+)
+
+diff --git a/drivers/misc/eeprom/optoe.c b/drivers/misc/eeprom/optoe.c
+index 22d2c0cd4..f3f9d89c4 100644
+--- a/drivers/misc/eeprom/optoe.c
++++ b/drivers/misc/eeprom/optoe.c
+@@ -528,6 +528,19 @@ static ssize_t optoe_eeprom_update_client(struct optoe_data *optoe,
+ 					page, ret);
+ 			return ret;
+ 		}
++	} else {
++		/* Set the page select byte to 0 if accessing the upper memory of page 0h (limit scope of fix to non-SFP modules) */
++		if (optoe->dev_class != TWO_ADDR && phy_offset >= OPTOE_PAGE_SIZE) {
++			page = 0;
++			ret = optoe_eeprom_write(optoe, client, &page,
++				OPTOE_PAGE_SELECT_REG, 1);
++			if (ret < 0) {
++				dev_err(&client->dev,
++					"Non-SFP write page register for page %d failed ret:%d!\n",
++						page, ret);
++				return ret;
++			}
++		}
```

Comment:
I'd recommend adding a `dev_warn` to say that SFP modules are not handled here.

Reply:
@saiarcot895 Adding a `dev_warn` will keep flooding the kernel logs since this function will be periodically called on SONiC as part of DOM monitoring.

Reply:
Ah ok, never mind then.

Reply:
I thought, Linux provides a function to only print a warning once or make it rate limited.

Reply:
Ah, you're right, there's `dev_warn_once` and `dev_warn_ratelimited`. Looks like the default ratelimit interval is 10 messages every 5 seconds, which may already be too much depending on how often DOM monitoring happens.

@mihirpat1 can you use `dev_warn_once` or `dev_warn_ratelimited`? For using a custom ratelimit, see https://github.com/torvalds/linux/blob/4dc1d1bec89864d8076e5ab314f86f46442bfb02/drivers/s390/cio/vfio_ccw_cp.c#L714-L732 as an example.

Reply:
@saiarcot895 @paulmenzel in the case of SFP, the offset in the range 128 to 256 in address 0xA2 is user writable

We don't have a use case to support that yet(SFP is very old transceiver form factor and not going to be used). This fix is needed when offset falls into that range. We can put a comment in the code in case anyone want to fix that later and test. IMO, we don't need a warning for this as you can see nobody even hit this issue or thought to fix this again for the same reason - We don't have SFPs that need user writable area to be written/read . 


![image](https://github.com/user-attachments/assets/e7f98ed9-2fc2-4608-8384-20a462ce45b4)